public/js/toc.js

```diff
@@ -0,0 +1,125 @@
+
+
+    //Generate regexp for different header levels
+    var lvl1Reg = new RegExp(/^#{1}/);
+    var lvl2Reg = new RegExp(/^#{2}/);
+    var lvl3Reg = new RegExp(/^#{3}/);
+    var lvl4Reg = new RegExp(/^#{4}/);
+    var lvl5Reg = new RegExp(/^#{5}/);
+    var lvl6Reg = new RegExp(/^#{6}/);
+
+    var header = {};
+    var len = this.lines.length;
+    for (i = 0; i<len; i++){
+      //sift down levels
+      if (this.lines[i].search(lvl6Reg) != -1){
+        header = {};
+        header.level = 6;
+        header.hText = this.lines[i].substring(6,this.lines[i].length);
+        header.lineNum = i;
+        header.idStr = "#toc"+header.lineNum; // Store line number information in ID. Rather than use some global lookup table
+        this.heads.push(header);
+
+      }
+      else if (this.lines[i].search(lvl5Reg) != -1){
+        header = {};
+        header.level = 5;
+        header.hText = this.lines[i].substring(5,this.lines[i].length);
+        header.lineNum = i;
+        header.idStr = "#toc"+header.lineNum;
+        this.heads.push(header);
+
+      }
+      else if (this.lines[i].search(lvl4Reg) != -1){
+        header = {};
+        header.level = 4;
+        header.hText = this.lines[i].substring(4,this.lines[i].length);
+        header.lineNum = i;
+        header.idStr = "#toc"+header.lineNum;
+        this.heads.push(header);
+
+      }
+      else if (this.lines[i].search(lvl3Reg) != -1){
+        header = {};
+        header.level = 3;
+        header.hText = this.lines[i].substring(3,this.lines[i].length);
+        header.lineNum = i;
+        header.idStr = "#toc"+header.lineNum;
+        this.heads.push(header);
+
+      }
+      else if (this.lines[i].search(lvl2Reg) != -1){
+        header = {};
+        header.level = 2;
+        header.hText = this.lines[i].substring(2,this.lines[i].length);
+        header.lineNum = i;
+        header.idStr = "#toc"+header.lineNum;
+        this.heads.push(header);
+
+      }
+      else if (this.lines[i].search(lvl1Reg) != -1){
+        header = {};
+        header.level = 1;
+        header.hText = this.lines[i].substring(1,this.lines[i].length);
+        header.lineNum = i;
+        header.idStr = "#toc"+header.lineNum;
+        this.heads.push(header);
+
+      }
+      }
+    };
+  /*
+  * This generates a header by appending a list to the root container with
+  *
+  */
+  this.generateHeaders = function(){
+    // empty out the toc  (crazy what kinds of stuff you can get away with)
+    $('#toc-root').empty();
+    this.parse();
+    var len = this.heads.length;
+    for(i = 0;i<len;i++){
+      if (this.heads[i].hText.length > 0){
+        //keeping the nested list in case we want to add expanding and contracting functionality later
+        $('#toc-root').append('<ul style = "list-style-type:none" class = "lvl" ><li><a  class = "list-group-item level level'+this.heads[i].level+'"  id = "'+this.heads[i].idStr+'">'+this.heads[i].hText+'</a></li></ul>');
```

Comment:
Could store the line number in a data attribute. Would avoid parsing the line number out of the id string later.